core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -299,48 +292,36 @@ class ReplicaFetcherThread(name: String,
     partition.truncateFullyAndStartAt(offset, isFuture = false)
   }
 
-  override def fetchEpochsFromLeader(partitions: Map[TopicPartition, EpochData]): Map[TopicPartition, EpochEndOffset] = {
-    def undefinedResponseMap(error: Errors,
-                             requestMap: Map[TopicPartition, EpochData]): Map[TopicPartition, EpochEndOffset] = {
-      requestMap.map { case (tp, _) =>
-        tp -> new EpochEndOffset(error, UNDEFINED_EPOCH, UNDEFINED_EPOCH_OFFSET)
-      }
+  override def fetchEpochEndOffsets(partitions: Map[TopicPartition, EpochData]): Map[TopicPartition, EpochEndOffset] = {
+
+    if (partitions.isEmpty) {
```

Comment:
I'm not sure whether this makes sense to have in the code anymore, I think it can never evaluated to true.

It is only called from `truncateToEpochEndOffsets` which is only called in `maybeTruncate()` like this:
```
if (partitionsWithEpochs.nonEmpty) {
      truncateToEpochEndOffsets(partitionsWithEpochs)
    }
```

Reply:
Fair enough. Sometimes it is difficult to strike a good balance good defensiveness and avoiding redundant validation. I think I will just get rid of the `nonEmpty` check.

Reply:
After looking at it, I think I prefer to leave it as is. The cost is small and it makes the code a little more resilient if we can avoid unnecessary assumptions about arguments, particularly when the logic is split across separate classes.